Spark 3.4: Add write options to override the compression properties of the table

jerqi opened the pull request:
### What changes were proposed in this pull request?
Flink has supported to use write option to override the compression properties of the table.
I refer to the pull request #6049 and then I add write options to override the compression properties of the table for Spark 3.4.

### Why are the changes needed?
First, there exist some old tables using `gzip` compression codec in the  production environment.  If we use `zstd` compression codec to rewrite the data, we can reduce the data volume and reduce the cost of storage.
Second, this pr is also meaning after we choose `zstd` as default compression codec.  Because we can choose different compression levels when we write the Iceberg data and rewrite the Iceberg data if we have this pr. 

### Does this PR introduce _any_ user-facing change?
Yes. 
This pr introduces the config option ` compression-codec`, `compression-level` and `compression-strategy`. (document added)
This pr introduces the Spark config properties `spark.sql.iceberg.write.compression-codec`,  `spark.sql.iceberg.write.compression-level` and `spark.sql.iceberg.write.compression-strategy`. (I will add the document after this pr is merged.)

### How was this patch tested?
Add a new ut and manual verification.
I use the ut to verify that the compression codec is correct.
I verify the compression config option is correct by hand.
<img width="514" alt="企业微信截图_93ca4fed-cd04-4315-819b-a56afdea106c" src="https://github.com/apache/iceberg/assets/8159038/3b8636a6-15f1-4ca9-9b5a-af3dfbccf6f8">

<img width="633" alt="企业微信截图_fdb11295-8800-4ccc-ab5a-ec0abf12e913" src="https://github.com/apache/iceberg/assets/8159038/29242681-5b1c-48e1-86c8-52b8fd898612">

jerqi commented:
@chenjunjiedada @ConeyLiu Could you help me review this pr?

chenjunjiedada commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -0,0 +1,180 @@
+    for (int i = 0; i < 1000; i++) {
+      expectedOrigin.add(new SimpleRecord(1, "hello world" + i));
+    }
+    Dataset<Row> df = spark.createDataFrame(expectedOrigin, SimpleRecord.class);
```
nit： blank line after block.

chenjunjiedada commented:
@jerqi Could you add some doc like in #6049 since `compression-level` and `compression-strategy` are targeted for different formats separately?  Also, it would be better to add a description of the default value and the available options. Others LGTM. Thanks for the contribution.

jerqi commented:
> @jerqi Could you add some doc like in #6049 since `compression-level` and `compression-strategy` are targeted for different formats separately? Also, it would be better to add a description of the default value and the available options. Others LGTM. Thanks for the contribution.

Thanks for your review. I have add the write option document for Spark. And I find that we don't have Spark properties document. I want to raise another pr to add it after this pr is merged.

jerqi commented:
@pvary Could you help me review this pr?

ConeyLiu commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -418,4 +418,64 @@ public String branch() {
+        .sessionConf(SparkSQLProperties.COMPRESSION_LEVEL)
+        .tableProperty(TableProperties.PARQUET_COMPRESSION_LEVEL)
+        .defaultValue(TableProperties.PARQUET_COMPRESSION_LEVEL_DEFAULT)
+        .parseOptional();
```
We can not get an optional value because we have set the default value, right?

jerqi replied:
Yes, good catch.

jerqi replied:
The default value is `null`, so we use the method `parseOptional`.

ConeyLiu commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -418,4 +418,64 @@ public String branch() {
+        .sessionConf(SparkSQLProperties.COMPRESSION_LEVEL)
+        .tableProperty(TableProperties.AVRO_COMPRESSION_LEVEL)
+        .defaultValue(TableProperties.AVRO_COMPRESSION_LEVEL_DEFAULT)
+        .parseOptional();
```
Sames here

jerqi replied:
Good catch! Flink have a similar issue. I will raise another pr to correct the code of Flink.

jerqi replied:
ditto.

ConeyLiu commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -0,0 +1,181 @@
+@RunWith(Parameterized.class)
+public class TestCompressionSettings {
+
+  private static final Configuration CONF = new Configuration();
```
It would be better to cluster as follows:
```
private static final
private final
private
```

jerqi replied:
Fixed.

ConeyLiu commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -0,0 +1,181 @@
+    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifestFiles.get(0), table.io())) {
+      DataFile file = reader.iterator().next();
+      InputFile inputFile = table.io().newInputFile(file.path().toString());
+      Assert.assertEquals(
```
Please follow the contribution guide using `AssertJ` instead. You can refer here: https://iceberg.apache.org/contribute/#assertj

jerqi replied:
Fixed.

ConeyLiu reviewed (approved):
+1

nastra commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -0,0 +1,182 @@
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestCompressionSettings {
```
The project is currently in the process of moving away from JUnit4, meaning that new tests should be written purely in Junit5 + using AssertJ assertions. See also https://iceberg.apache.org/contribute/#testing for some additional info

jerqi replied:
Thanks for your review. I got it. I have migrated the test to the Junit 5 framework.

aokolnychyi reviewed (commented):
Do we also have to cover other places that use `SparkFileWriterFactory`?
For instance, `SparkPositionDeltaWrite`?

jerqi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:
```diff
@@ -63,4 +63,9 @@ private SparkSQLProperties() {}
   public static final String WAP_BRANCH = "spark.wap.branch";
+
+  // Controls write compress options
+  public static final String COMPRESSION_CODEC = "spark.sql.iceberg.write.compression-codec";
```
Fixed.

jerqi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkFileWriterFactory.java:
```diff
@@ -76,6 +79,12 @@ class SparkFileWriterFactory extends BaseFileWriterFactory<InternalRow> {
     this.dataSparkType = dataSparkType;
     this.equalityDeleteSparkType = equalityDeleteSparkType;
     this.positionDeleteSparkType = positionDeleteSparkType;
+
```
Fixed.

jerqi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -178,6 +185,32 @@ private WriterFactory createWriterFactory() {
     // broadcast the table metadata as the writer factory will be sent to executors
     Broadcast<Table> tableBroadcast =
         sparkContext.broadcast(SerializableTableWithSize.copyOf(table));
+    Map<String, String> writeProperties = Maps.newHashMap();
```
Fixed.

jerqi commented:
> Do we also have to cover other places that use `SparkFileWriterFactory`? For instance, `SparkPositionDeltaWrite`?

Thanks for you review, my mistake. I will add the modification of the code and add the ut.

jerqi commented on spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -0,0 +1,222 @@
+          .isEqualToIgnoringCase(properties.get(COMPRESSION_CODEC));
+    }
+
+    if (PARQUET.equals(format)) {
```
I find that we can't set table property for the action `RewritePositionDeletes`.  The action `RewritePositionDeletes` will always use parquet as the format.

jerqi replied:
<img width="726" alt="企业微信截图_491440a5-40ce-404f-86c0-aec1addfe029" src="https://github.com/apache/iceberg/assets/8159038/ac998508-c3a6-4918-8038-0cfac9fa18d1">
In my ut, I set the table property `DEFAULT_FILE_FORMAT` is orc, but the table properties is empty.

aokolnychyi replied:
That seems like an issue worth looking into after this change.

cc @szehon-ho.

jerqi replied:
OK, I will investigate it.

jerqi replied:
I find the root cause of this case.
The class `BaseMetadataTable` override the method `properties`
```
  @Override
  public Map<String, String> properties() {
    return ImmutableMap.of();
  }
```
Option 1: 
we can modify the class `BaseMetadataTable`.
https://github.com/apache/iceberg/pull/8428

Option2:
We can modify the class `PositionDeletesTable`
https://github.com/apache/iceberg/pull/8429

I prefer option 1. I feel the properties of  meta data table should respect the ones of  base table. We also don't have ways to modify the properites of  meta data table. @aokolnychyi @szehon-ho WDYT?

szehon-ho replied:
Hi, yea I am ok with option 1 as well,  @aokolnychyi what do you think?

jerqi commented:
> > Do we also have to cover other places that use `SparkFileWriterFactory`? For instance, `SparkPositionDeltaWrite`?
> 
> Thanks for you review, my mistake. I will add the modification of the code and add the ut.

@aokolnychyi I have addressed the comments.

jerqi commented:
@nastra @aokolnychyi Gently ping. Sorry to bother you. Could you have another look if you have time?

aokolnychyi commented:
Let me take another look.

aokolnychyi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:
```diff
@@ -0,0 +1,66 @@
+  private WritePropertiesUtil() {}
+
+  public static Map<String, String> writeProperties(FileFormat format, SparkWriteConf writeConf) {
+    Map<String, String> writeProperties = Maps.newHashMap();
```
Optional: What about an empty line before the switch?

jerqi replied:
OK, I got it.

aokolnychyi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.spark.SparkWriteConf;
+
+public class WritePropertiesUtil {
```
I wonder whether it is worth adding a new util class in this case. Do we expect more related logic? I feel like we could just use `SparkWriteConf` and offer `writeProperties(FileFormat format)` method there, potentially making the newly added method like `parquetCompressionLevel` private. Any thoughts?

jerqi replied:
Good point. Thanks for your suggestion.

aokolnychyi commented:
This change looks good to me. I am not sure about adding a new utility class vs just using `SparkWriteConf`.
@jerqi, what do you think?

jerqi commented:
> This change looks good to me. I am not sure about adding a new utility class vs just using `SparkWriteConf`. @jerqi, what do you think?

The second solution seems more elegant. Thanks for your suggestion.  I have followed the suggestion.

aokolnychyi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:
```diff
@@ -178,6 +178,7 @@ private WriterFactory createWriterFactory() {
     // broadcast the table metadata as the writer factory will be sent to executors
     Broadcast<Table> tableBroadcast =
         sparkContext.broadcast(SerializableTableWithSize.copyOf(table));
+
```
Minor: Is this change deliberate?

aokolnychyi reviewed (approved):
Looks good to me. Left a few suggestions.

aokolnychyi commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:
```diff
@@ -104,6 +104,7 @@ class SparkPositionDeltaWrite implements DeltaWrite, RequiresDistributionAndOrde
   private final Context context;
 
   private boolean cleanupOnAbort = true;
+  private final Map<String, String> writeProperties;
```
Minor: This variable should be co-located with other final variables above. It is super minor but let's do this in a follow-up.

jerqi replied:
OK, I have raised a follow-up pr https://github.com/apache/iceberg/pull/8421.

aokolnychyi commented:
Thanks, @jerqi! Thanks for reviewing, @ConeyLiu @chenjunjiedada.

jerqi commented:
Thanks @aokolnychyi ! Code master. Thanks @ConeyLiu @chenjunjiedada @nastra , too.

szehon-ho commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -418,4 +424,96 @@ public String branch() {
+        writeProperties.put(PARQUET_COMPRESSION, parquetCompressionCodec());
+        String parquetCompressionLevel = parquetCompressionLevel();
+        if (parquetCompressionLevel != null) {
+          writeProperties.put(PARQUET_COMPRESSION_LEVEL, parquetCompressionLevel);
```
Hi guys, while working on #8158 , I noticed the test failing and that this is missing delete file compression override.  I think the TestCompression delete file compression check passes by chance, I have a fix over there for it: https://github.com/apache/iceberg/pull/8299/commits/43e3cabb21964c7e2a9ebe424dcf4d1f03b8006c

jerqi replied:
Sorry for this mistake. I will check how to valid the correctness of the case.
